src/cpp/exe/presolve/main.cpp

```diff
@@ -0,0 +1,179 @@
+
+    // Parse structure and get candidates' indices
+    logger->display_message("Reading " + structure_path.string());
+    const CouplingMap full_couplings = CouplingMapGenerator::BuildInput(structure_path,
+                                                                        logger.get(),
+                                                                        "Presolve");
+
+    // TODO Add option to keep or discard the full versions
+    // TODO All filesystem operations can be done in a separated part I guess
+    // Creates new folder to move full problems
+    logger->display_message("Creating " + full_dir.string());
+    mkdir(full_dir);
+
+    // Move full structure to 'full' folder
+    std::filesystem::rename(structure_path, full_dir / options.STRUCTURE_FILE);
+
+    // ** Main part : creates coupling map for reduced problems **
+    CouplingMap reduced_couplings = full_couplings;
+
+    const size_t nb_prob_total{full_couplings.size() - 1};
+    for (int nb_prob{0}; const auto& [filename, var_map]: full_couplings)
+    {
+        if (filename == options.MASTER_NAME) [[unlikely]]
+        {
+            // Keep master indices untouched
+            // and only try to reduce the subproblems
+            continue;
+        }
+
+        // Get all candidate indices in a sorted array
+        std::vector<int> colind(var_map.size());
+        std::transform(var_map.cbegin(),
+                       var_map.cend(),
+                       colind.begin(),
+                       [](const auto pair) { return pair.second; });
+        std::sort(colind.begin(), colind.end());
+
+        // Read full problem
+        const std::filesystem::path subproblem_path = input_root_dir / filename;
+
+        // TODO See about the keeprows thing
+        solver.read_prob_mps(subproblem_path);
+
+        // Keep the solver from removing candidate indices from subproblem
+        solver.mark_indices_to_keep_presolve(0, colind.size(), nullptr, colind.data());
+
+        solver.presolve_only();
+
+        // Move full subproblem to 'full' folder
+        // TODO Add option to keep or discard the full versions
+        // TODO All filesystem operations can be done in a separated part I guess
+        std::filesystem::rename(subproblem_path, full_dir / filename);
+
+        logger->display_message(
+          fmt::format("Subproblem '{}' reduced: {} / {}", filename, ++nb_prob, nb_prob_total));
+
+        // Write reduced problem MPS
+        solver.write_prob_mps(subproblem_path);
+
+        // Create a map [full_idx] -> reduced_idx for candidate indices
+        const auto full2reduced = get_presolve_map(solver, colind);
+
+        for (const auto& [var_name, idx]: var_map)
+        {
+            reduced_couplings.at(filename).at(var_name) = full2reduced.at(idx);
+        }
+    }
+
+    return reduced_couplings;
+}
+
+void write_reduced_problems(const CouplingMap& reduced_couplings,
```

Comment:
```suggestion
void write_reduced_couplings(const CouplingMap& reduced_couplings,
```
`write_reduced_problems` is confusing, it looks like you are writing the mps files